### PR TITLE
EOS-19917 CORTX-HARE: Implement test interface for Hare

### PR DESCRIPTION
**Pull request by SwapnilGaonkar7**
Signed-off-by: Swapnil Gaonkar <swapnil.gaonkar@seagate.com>

### CODE REVIEW

File: provisioning/setup.yaml [hunk 1]
```diff
@@ -10,7 +10,7 @@ hare:
     args: --config $URL --file '/var/lib/hare/cluster.yaml'
   test:
     cmd: /opt/seagate/cortx/hare/bin/hare_setup test
-    args: --config $URL
+    args: --config $URL --plan $TYPE --param $PARAM_URL
```
Comment (nitin-seagate) [inline provisioning/setup.yaml:13]: We never proposed `--param $PARAM_URL`
Please provide context for this usage?

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: provisioning/setup.yaml (hunk 1 above)
Comment (SwapnilGaonkar7) [reply provisioning/setup.yaml:13]: @nitin-seagate [This](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/221642873/CORTX+Factory+Manufacturing+Process) page specify following command. It is also being discussed in integration meeting.
`cortx_cli test run --plan {sanity|regression|full|performance|scalability} --param <TEST_INPUT_URL>`

Comment (knekrasov) [reply provisioning/setup.yaml:13]: @SwapnilGaonkar7 if Nitin has concerns, maybe it is too risky to merge this feature now? Won't this change break deployment (say, if setup.yaml parser fails..)

Comment (SwapnilGaonkar7) [reply provisioning/setup.yaml:13]: @knekrasov  Yes, we are not going to merge this until @nitin-seagate confirms.

File: provisioning/miniprov/hare_mp/main.py [hunk 2]
```diff
@@ -232,6 +240,28 @@ def test(args):
...
 
+def test_IVT(args):
+    try:
+        rc = 0
+        path_to_cdf = args.file[0]
+
+        logging.info('Running test plan: ' + str(args.plan[0].value))
+        # TODO We need to handle plan type and execute test cases accordingly
+        if not is_cluster_running():
+            logging.error('Cluster is not running. Cluster must be running '
```
Comment (knekrasov) [inline provisioning/miniprov/hare_mp/main.py:251]: This is not so obvious to me. Is it required by any specification?

Or, speaking wider, is there any requirement in which state the system will be when `test` is invoked? And in which state the system must be left when `test` stage finishes?

Comment (SwapnilGaonkar7) [reply provisioning/miniprov/hare_mp/main.py:251]: @knekrasov It is expected that cluster will be already running when 'test' gets called.
We should not start/stop cluster during test stage

Comment (mssawant) [reply provisioning/miniprov/hare_mp/main.py:251]: So now `test_IVT` command is replacing `test` command. I believe `test()` needs to be invoked for plan `Sanity`, in that case please add that call as well.
You can have a dictionary mapping the plan to a function. Have a test function per plan, e.g. rename test() -> test_sanity(). Similarly you can define test_regression, test_functionality, etc.

Comment (SwapnilGaonkar7) [reply provisioning/miniprov/hare_mp/main.py:251]: @mssawant Functions for specific plan needs to go in test rpm. So there will be file which will be installed as part of hare test rpm which will receive command from our hare_setup and will decide which functions will be called based on plan selected while executing IVT.
Let me know your thoughts.

Comment (mssawant) [reply provisioning/miniprov/hare_mp/main.py:251]: Don't we atleast need dummy functions or a check atleast in this case. Because, the command (hare_setup test) is part of the cortx-hare rpm but the implmentation is part of a different rpm. Now, use can see the command and when tries to execute it will fail for the given plan because test rpm is not installed.

Comment (knekrasov) [reply provisioning/miniprov/hare_mp/main.py:251]: This problem with several RPMs can be resolved as follows:
1. Main Hare RPM provides this CLI to invoke tests. It parses the CLI parameters and extracts the "plan" name
2. All the tests are pytest-driven.
3. The tests that correspond to a certain plan can be either stored in separate folders (folder name corresponds to plan name somehow) or the plan is mapped to a [mark within the test](https://docs.pytest.org/en/6.2.x/example/markers.html).
4. Test RPM does nothing except deploying some _extended_ set of test files into the location which our `test` CLI is already aware of.

So in a nutshell `test` CLI will always invoke pytest within a given folder. The folder may be empty if test RPM is not installed.

Comment (SwapnilGaonkar7) [reply provisioning/miniprov/hare_mp/main.py:251]: @mssawant We will add check similar to below once we add code for accessing hare rpm functionality.

> if checkRpm('cortx-hare-test') != 0:
            logging.error('\'cortx-hare-test\' is not installed')
            exit(-1)

Let me know your inputs

Comment (SwapnilGaonkar7) [reply provisioning/miniprov/hare_mp/main.py:251]: @mssawant @knekrasov I have also raised PR for generating test rpm.
Once test rpm is installed, following commonad can be invoked from this interface
`/opt/seagate/cortx/hare/libexec/hare_test execute --plan sanity`

File: provisioning/setup.yaml (hunk 1 above)
Comment (nitin-seagate) [reply provisioning/setup.yaml:13]: @SwapnilGaonkar7 The cortx CLI command is not one to one mapping to Mini-Provisioner interfaces...
For reference please check this section:
https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/85000402/Mini-Provisioner+ConfStore+Keys#Test-Component-ConfStore%3A
Please define which keys you need to run Test RPM for Hare

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: provisioning/setup.yaml (hunk 1 above)
Comment (SwapnilGaonkar7) [reply provisioning/setup.yaml:13]: @nitin-seagate Currently our test section do not use keys from confstore. We will update if that changes.
Please confirm if above setup.yaml changes need to be altered.

**Comment by mssawant** [comment]
@SwapnilGaonkar7, kindly squash or rebase the branch with latest main.

File: provisioning/setup.yaml (hunk 1 above)
Comment (nitin-seagate) [reply provisioning/setup.yaml:13]: @SwapnilGaonkar7  Please remove `-param $PARAM_URL` for now if you are not using any test keys

Comment (SwapnilGaonkar7) [reply provisioning/setup.yaml:13]: Done

**Comment by SwapnilGaonkar7** [comment]
@mssawant @knekrasov I have addresses Nitin's comment and also done rebase.
Please have a look

File: provisioning/miniprov/hare_mp/main.py (hunk 2 above)
Comment (knekrasov) [reply provisioning/miniprov/hare_mp/main.py:251]: As discussed in that PR `hare_test` doesn't seem to make sense

**Review by knekrasov — APPROVED** [review]

**Review by mssawant — APPROVED** [review]
